### PR TITLE
Disable gravity compensation

### PR DESCRIPTION
**Pull request by cgokmen**
Here we disable gravity compensation but leave it as an option for fixed-base robots. It is fully removed for floating base robots.

### CODE REVIEW

File: omnigibson/controllers/joint_controller.py [hunk 1]
```diff
@@ -42,6 +46,8 @@ def __init__(
         kp=None,
         damping_ratio=None,
         use_impedances=False,
+        use_gravity_compensation=False,
+        use_cc_compensation=False,
```
Comment (cremebrule) [inline omnigibson/controllers/joint_controller.py:50]: i think we still need CC compensation (And this value is accurate IIRC) -- so this should be True by default

File: omnigibson/controllers/joint_controller.py [hunk 2]
```diff
@@ -74,6 +80,10 @@ def __init__(
                 damping ratio applied to the joint controller. If None, a default value will be used.
             use_impedances (bool): If True, will use impedances via the mass matrix to modify the desired efforts
                 applied
+            use_gravity_compensation (bool): If True, will add gravity compensation to the computed efforts. This is
+                an experimental feature that only works on fixed base robots. We do not recommend enabling this.
+            use_cc_compensation (bool): If True, will add Coriolis / centrifugal compensation to the computed efforts
+                This is an experimental feature. We do not recommend enabling this.
```
Comment (cremebrule) [inline omnigibson/controllers/joint_controller.py:86]: fix -- i think we should default to including this

File: omnigibson/controllers/joint_controller.py [hunk 3]
```diff
@@ -99,6 +109,20 @@ def __init__(
...
+        # Warn the user about gravity compensation and Coriolis / centrifugal compensation being experimental.
+        if self._use_gravity_compensation:
+            log.warning(
+                "JointController is using gravity compensation. This is an experimental feature that only works on "
+                "fixed base robots. We do not recommend enabling this."
+            )
+        if self._use_cc_compensation:
+            log.warning(
+                "JointController is using Coriolis / centrifugal compensation. This is an experimental feature. We do "
+                "not recommend enabling this."
```
Comment (cremebrule) [inline omnigibson/controllers/joint_controller.py:124]: fix

File: omnigibson/controllers/osc_controller.py [hunk 4]
```diff
@@ -69,6 +69,8 @@ def __init__(
         mode="pose_delta_ori",
         decouple_pos_ori=False,
         workspace_pose_limiter=None,
+        use_gravity_compensation=False,
+        use_cc_compensation=False,
```
Comment (cremebrule) [inline omnigibson/controllers/osc_controller.py:73]: same here

File: omnigibson/objects/controllable_object.py [hunk 5]
```diff
@@ -189,6 +188,18 @@ def load(self, scene):
...
+    def _post_load(self):
+        # Call super first
+        super()._post_load()
+
+        # For controllable objects, we disable gravity of all links that are not the base link. This
+        # is because we cannot accurately apply gravity compensation in the absence of a working
+        # generalized gravity force computation. This may have some side effects on the measured
+        # torque on each of these links, but it provides a greatly improved joint control behavior.
+        for link in self._links.values():
+            if link is not self.root_link:
```
Comment (cremebrule) [inline omnigibson/objects/controllable_object.py:200]: this is quite elegant, but may be problematic depending on the robot. For example, isn't Tiago's root link technically x_base_footprint? And a robot's root link may not actually be the highest mass link

A very hacky, but perhaps more accurate heuristic could be to only select the largest mass link to be gravity enabled.

File: omnigibson/robots/robot_base.py [hunk 6]
```diff
@@ -129,7 +129,6 @@ def __init__(
 
         # Initialize internal attributes that will be loaded later
         self._sensors = None  # e.g.: scan sensor, vision sensor
-        self._dummy = None  # Dummy version of the robot w/ fixed base for computing generalized gravity forces
```
Comment (cremebrule) [inline omnigibson/robots/robot_base.py:132]: BLESS

File: omnigibson/utils/ui_utils.py [hunk 7]
```diff
@@ -590,7 +590,7 @@ def __init__(self, robot):
                 "command_dim": controller.command_dim,
             }
             idx += controller.command_dim
-            for i in controller.dof_idx:
+            for i in controller.dof_idx.tolist():
```
Comment (cremebrule) [inline omnigibson/utils/ui_utils.py:593]: is this because it could be shape (1, N) instead of (N,)?

**Review by cremebrule — COMMENTED** [review]
overall LGTM, a couple of comments

File: omnigibson/utils/ui_utils.py (hunk 7 above)
Comment (cgokmen) [reply omnigibson/utils/ui_utils.py:593]: It's just an error that happened in the torch conversion. `i` needs to be an int and not a tensor.

File: omnigibson/robots/robot_base.py (hunk 6 above)
Comment (cgokmen) [reply omnigibson/robots/robot_base.py:132]: :))))))))

File: omnigibson/objects/controllable_object.py (hunk 5 above)
Comment (cgokmen) [reply omnigibson/objects/controllable_object.py]: I have implemented our discussesd solution.

File: omnigibson/objects/controllable_object.py [hunk 8]
```diff
@@ -189,6 +189,34 @@ def load(self, scene):
...
+        # Note that we do NOT disable gravity for links that are fixed to the base link, as these links
+        # are typically where most of the downward force on the robot is applied. Disabling gravity
+        # for these links would result in the robot floating in the air easily. Also note that here
+        # we use the base link footprint which takes into account the presence of virtual joints.
+
+        # Find all the links that are accessible from the base link footprint via a chain of fixed
+        # joints. We will disable gravity for all links that are not in this set.
+        articulation_tree = self.articulation_tree
+        base_footprint = self.base_footprint_link_name
+        is_edge_fixed = lambda f, t: articulation_tree[f][t]["joint_type"] == "FixedJoint"
```
Comment (cremebrule) [inline omnigibson/objects/controllable_object.py:209]: use JointType class from utils.constants

**Review by cremebrule — APPROVED** [review]

**Pull request merged**